sharktank/tests/ops/pipeline_parallelized_test.py

```diff
@@ -0,0 +1,350 @@
+        assert actual_result.pinned == expected_result.pinned
+        assert all(d_act == d_exp for d_act, d_exp in zip(actual_result.devices, expected_result.devices))
+
+    def testReshardLikeShardedToShared(self):
+        tensor = torch.rand(5, 6, dtype=torch.float32)
+        shard_dim = 1
+        shard_count = 3
+        expected_result = ops.reshard_split(tensor, dim=shard_dim, count=shard_count).clone(pinned=False)
+        target = ops.reshard_split(tensor, dim=shard_dim, count=shard_count).clone(devices=tuple(2*i for i in range(shard_count)), pinned=True)
+
+        actual_result = ops.reshard_like(expected_result, target)
+        assert expected_result.is_deep_equal(actual_result)
+        assert actual_result.pinned == target.pinned
+        assert all(d_act == d_targ for d_act, d_targ in zip(actual_result.devices, target.devices))
+
+    def testReshardLikeReplicatedToReplicated(self):
+        tensor = torch.rand(4, 5, 6, dtype=torch.float32)
+        shard_count = 2
+        input_tensor = ops.replicate(tensor, count=shard_count).clone(devices=tuple(range(shard_count)), pinned=False)
+        target = ops.replicate(tensor, count=shard_count).clone(devices=tuple(2*i for i in range(shard_count)), pinned=True)
+        
+        actual_result = ops.reshard_like(input_tensor, target)
+        assert input_tensor.is_deep_equal(actual_result)
+        assert actual_result.pinned == target.pinned
+        assert all(d_act == d_targ for d_act, d_targ in zip(actual_result.devices, target.devices))
+
+    def testReshardLikeReplicatedToSharded(self):
+        tensor = torch.rand(4, 5, 6, dtype=torch.float32)
+        shard_dim = 2
+        shard_count = 3
+        expected_result = ops.reshard_split(tensor, dim=shard_dim, count=shard_count).clone(devices=tuple(2*i for i in range(shard_count)), pinned=True)
+        replicated_tensor = ops.replicate(tensor, count=shard_count).clone(devices=tuple(range(shard_count)), pinned=False)
+        
+        actual_result = ops.reshard_like(replicated_tensor, expected_result)
+        assert expected_result.is_deep_equal(actual_result)
+        assert actual_result.pinned == expected_result.pinned
+        assert all(d_act == d_exp for d_act, d_exp in zip(actual_result.devices, expected_result.devices))
+
+class TransposeTest(unittest.TestCase):
+    def testUnpinnedTranspose(self):
+        shard_count = 4
+        shard_shape = [3, 4]
+        devices = tuple(2+i for i in range(shard_count))
+        shards = [torch.rand(shard_shape, dtype=torch.float32) for _ in range(shard_count)]
+        pre = SplitPrimitiveTensor(shard_dim=1, ts=shards, devices=devices, pinned=False)
+
+        post = pre.T
+        assert all(d_pre == d_post for d_pre, d_post in zip(pre.devices, post.devices))
+        # TODO: post gets pinned since resulting ShardedTensor is made with torch.Tensor shards which are assumed to always be pinned
+        # assert post.pinned == pre.pinned
```

Comment:
This may be an issue

Reply:
Meh, pinning is more for noting information. Comparing this is not really necessary as its not the intention of correctness.